crates/modelardb_server/src/configuration.rs

```diff
@@ -0,0 +1,154 @@
+/* Copyright 2023 The ModelarDB Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Management of the system's configuration, including the server mode, and the amount of
+//! reserved memory for uncompressed and compressed data.
+
+use std::sync::Arc;
+use modelardb_common::errors::ModelarDbError;
+use tokio::sync::RwLock;
+
+use crate::storage::StorageEngine;
+use crate::ServerMode;
+
+/// Store's the system's configuration and provides functionality for updating the configuration.
+#[derive(Clone)]
+pub struct ConfigurationManager {
+    /// The mode of the server used to determine the behaviour when modifying the remote object store.
+    server_mode: ServerMode,
+    /// Amount of memory to reserve for storing uncompressed data buffers.
+    uncompressed_reserved_memory_in_bytes: usize,
+    /// Amount of memory to reserve for storing compressed data buffers.
+    compressed_reserved_memory_in_bytes: usize,
+}
+
+impl ConfigurationManager {
+    pub fn new(server_mode: ServerMode) -> Self {
+        Self {
+            server_mode,
+            uncompressed_reserved_memory_in_bytes: 512 * 1024 * 1024, // 512 MiB
+            compressed_reserved_memory_in_bytes: 512 * 1024 * 1024,   // 512 MiB
+        }
+    }
+
+    pub(crate) fn server_mode(&self) -> &ServerMode {
+        &self.server_mode
+    }
+
+    pub(crate) fn uncompressed_reserved_memory_in_bytes(&self) -> &usize {
+        &self.uncompressed_reserved_memory_in_bytes
+    }
+
+    /// Set the new value and update the uncompressed remaining reserved memory in the storage engine.
+    pub(crate) async fn set_uncompressed_reserved_memory_in_bytes(
+        &mut self,
+        new_uncompressed_reserved_memory_in_bytes: usize,
+        storage_engine: Arc<RwLock<StorageEngine>>,
+    ) {
+        let value_change = new_uncompressed_reserved_memory_in_bytes as isize
```

Comment:
I have kept the current structure to avoid having to pass both the new value and the old value since the uncompressed data manager only keeps track of the current remaining reserved memory and not the total. I do understand your point of having everything in one place but I think it makes sense to perform all checks (especially if more checks are added in the future) in the configuration manager.

I have added a comment to explain why we calculate `value_change`.